### PR TITLE
Transform classic loops to range-based for loops in module recognition

### PR DESCRIPTION
**Pull request by SunBlack**
Changes are based on the result of run-clang-tidy -header-filter='.*' -checks='-*,modernize-loop-convert' -fix

### CODE REVIEW

File: recognition/include/pcl/recognition/impl/implicit_shape_model.hpp [hunk 1]
```diff
@@ -404,8 +404,8 @@ pcl::features::ISMModel::saveModelToFile (std::string& file_name)
   for (unsigned int i_cluster = 0; i_cluster < number_of_clusters_; i_cluster++)
   {
     output_file << static_cast<unsigned int> (clusters_[i_cluster].size ()) << " ";
-    for (unsigned int i_visual_word = 0; i_visual_word < static_cast<unsigned int> (clusters_[i_cluster].size ()); i_visual_word++)
-      output_file << clusters_[i_cluster][i_visual_word] << " ";
+    for (const unsigned int &i_visual_word : clusters_[i_cluster])
```
Comment (taketwo) [inline recognition/include/pcl/recognition/impl/implicit_shape_model.hpp]: Minor: this is not an index variable anymore, so <strike>i_</strike>`visual_word` is more appropriate name.

File: recognition/include/pcl/recognition/ransac_based/rigid_transform_space.h [hunk 2]
```diff
@@ -229,20 +229,20 @@ namespace pcl
           int max_num_transforms = 0;
 
           // For each full leaf
-          for ( std::vector<CellOctree::Node*>::const_iterator leaf = full_leaves.begin () ; leaf != full_leaves.end () ; ++leaf )
+          for (const auto &full_leave : full_leaves)
```
Comment (taketwo) [inline recognition/include/pcl/recognition/ransac_based/rigid_transform_space.h]: Minor: singular from "leaves" is "leaf".

File: recognition/src/face_detection/rf_face_detector_trainer.cpp [hunk 3]
```diff
@@ -369,16 +369,16 @@ void pcl::RFFaceDetectorTrainer::detectFaces()
           std::vector<NodeType> leaves;
           dfe.evaluate (forest_, fhda, rse, eval_examples, 0, leaves);
 
-          for (size_t l = 0; l < leaves.size (); l++)
+          for (const auto &leave : leaves)
```
Comment (taketwo) [inline recognition/src/face_detection/rf_face_detector_trainer.cpp]: leaf

File: recognition/src/ransac_based/model_library.cpp [hunk 4]
```diff
@@ -130,18 +130,18 @@ ModelLibrary::addModel (const PointCloudIn& points, const PointCloudN& normals,
   int num_of_pairs = 0;
 
   // Run through all full leaves
-  for ( vector<ORROctree::Node*>::const_iterator leaf1 = full_leaves.begin () ; leaf1 != full_leaves.end () ; ++leaf1 )
+  for (const auto &full_leave : full_leaves)
```
Comment (taketwo) [inline recognition/src/ransac_based/model_library.cpp]: leaf

File: recognition/src/ransac_based/model_library.cpp [hunk 5]
```diff
@@ -130,18 +130,18 @@ ModelLibrary::addModel (const PointCloudIn& points, const PointCloudN& normals,
...
   {
-    const ORROctree::Node::Data* node_data1 = (*leaf1)->getData ();
+    const ORROctree::Node::Data* node_data1 = full_leave->getData ();
 
     // Get all full leaves at the right distance to the current leaf
     inter_leaves.clear ();
     octree.getFullLeavesIntersectedBySphere (node_data1->getPoint (), pair_width_, inter_leaves);
 
-    for ( list<ORROctree::Node*>::iterator leaf2 = inter_leaves.begin () ; leaf2 != inter_leaves.end () ; ++leaf2 )
+    for (const auto &inter_leave : inter_leaves)
```
Comment (taketwo) [inline recognition/src/ransac_based/model_library.cpp]: leaf

File: recognition/src/ransac_based/obj_rec_ransac.cpp [hunk 6]
```diff
@@ -321,13 +321,13 @@ pcl::recognition::ObjRecRANSAC::groupHypotheses(list<HypothesisBase>& hypotheses
   float transformed_point[3];
 
   // Add all rigid transforms to the discrete rigid transform space
-  for ( list<HypothesisBase>::iterator hypo_it = hypotheses.begin () ; hypo_it != hypotheses.end () ; ++hypo_it )
+  for (const auto &hypothese : hypotheses)
```
Comment (taketwo) [inline recognition/src/ransac_based/obj_rec_ransac.cpp]: hypothesis

File: recognition/src/ransac_based/obj_rec_ransac.cpp [hunk 7]
```diff
@@ -579,29 +579,29 @@ pcl::recognition::ObjRecRANSAC::filterGraphOfConflictingHypotheses (ORRGraph<Hyp
...
 
   // Compute the penalty for each graph node
-  for ( vector<ORRGraph<Hypothesis*>::Node*>::iterator it = nodes.begin () ; it != nodes.end () ; ++it )
+  for (auto &node : nodes)
   {
     size_t num_of_explained = 0;
 
     // Accumulate the number of pixels the neighbors are explaining
-    for ( set<ORRGraph<Hypothesis*>::Node*>::const_iterator neigh = (*it)->getNeighbors ().begin () ; neigh != (*it)->getNeighbors ().end () ; ++neigh )
+    for ( set<ORRGraph<Hypothesis*>::Node*>::const_iterator neigh = node->getNeighbors ().begin () ; neigh != node->getNeighbors ().end () ; ++neigh )
```
Comment (taketwo) [inline recognition/src/ransac_based/obj_rec_ransac.cpp]: Can this be converted to range-for as well?

File: recognition/src/ransac_based/obj_rec_ransac.cpp [hunk 8]
```diff
@@ -624,10 +624,10 @@ pcl::recognition::ObjRecRANSAC::testHypothesis (Hypothesis* hypothesis, int& mat
   float transformed_point[3];
 
   // The match/penalty loop
-  for ( std::vector<ORROctree::Node*>::const_iterator leaf_it = full_model_leaves.begin () ; leaf_it != full_model_leaves.end () ; ++leaf_it )
+  for (const auto &full_model_leave : full_model_leaves)
```
Comment (taketwo) [inline recognition/src/ransac_based/obj_rec_ransac.cpp]: leaf

File: recognition/src/ransac_based/obj_rec_ransac.cpp [hunk 9]
```diff
@@ -661,10 +661,10 @@ pcl::recognition::ObjRecRANSAC::testHypothesisNormalBased (Hypothesis* hypothesi
...
```
Comment (taketwo) [inline recognition/src/ransac_based/obj_rec_ransac.cpp]: leaf

File: recognition/src/ransac_based/orr_octree.cpp [hunk 10]
```diff
@@ -187,8 +187,8 @@ pcl::recognition::ORROctree::build (const PointCloudIn& points, float voxel_size
   else
   {
     // Iterate over all full leaves and average points
-    for ( vector<ORROctree::Node*>::iterator it = full_leaves_.begin() ; it != full_leaves_.end() ; ++it )
-      (*it)->getData ()->computeAveragePoint ();
+    for (const auto &full_leave : full_leaves_)
```
Comment (taketwo) [inline recognition/src/ransac_based/orr_octree.cpp]: leaf

File: recognition/src/ransac_based/orr_octree_zprojection.cpp [hunk 11]
```diff
@@ -110,7 +110,7 @@ pcl::recognition::ORROctreeZProjection::build (const ORROctree& input, float eps
   full_leaves_bounds[2] = std::numeric_limits<float>::infinity();
   full_leaves_bounds[3] = -std::numeric_limits<float>::infinity();
 
-  for (const auto& leave : full_leaves)
+  for (const auto &leave : full_leaves)
```
Comment (taketwo) [inline recognition/src/ransac_based/orr_octree_zprojection.cpp]: leaf

File: recognition/src/ransac_based/orr_octree_zprojection.cpp [hunk 12]
```diff
@@ -152,9 +152,9 @@ pcl::recognition::ORROctreeZProjection::build (const ORROctree& input, float eps
   int pixel_id = 0;
 
   // Project the octree full leaves onto the xy-plane
-  for (auto fl_it = full_leaves.cbegin () ; fl_it != full_leaves.cend () ; ++fl_it )
+  for (const auto &full_leave : full_leaves)
```
Comment (taketwo) [inline recognition/src/ransac_based/orr_octree_zprojection.cpp]: leaf

File: recognition/include/pcl/recognition/ransac_based/rigid_transform_space.h (hunk 2 above)
Comment (SunBlack) [reply recognition/include/pcl/recognition/ransac_based/rigid_transform_space.h]: Clang automatically only removes the `s` to switch to singular ;).

Comment (taketwo) [reply recognition/include/pcl/recognition/ransac_based/rigid_transform_space.h]: Shame on you, clang!

File: recognition/src/face_detection/face_detector_data_provider.cpp [hunk 13]
```diff
@@ -76,24 +76,19 @@ void pcl::face_detection::FaceDetectorDataProvider<FeatureType, DataSet, LabelTy
     }
   }
 
-  for (size_t i = 0; i < files.size (); i++)
+  for (const auto &filename : files)
   {
-    std::stringstream filestream;
-    filestream << data_dir << "/" << files[i];
-    std::string file = filestream.str ();
+    std::string file = data_dir + "/" + filename;
```
Comment (SergioRAgostinho) [inline recognition/src/face_detection/face_detector_data_provider.cpp]: Isn't clang tidy going to complain about using strings for a single char?

File: recognition/src/face_detection/face_detector_data_provider.cpp [hunk 14]
```diff
@@ -76,24 +76,19 @@ void pcl::face_detection::FaceDetectorDataProvider<FeatureType, DataSet, LabelTy
...
+    std::string pose_file (filename);
     boost::replace_all (pose_file, ".pcd", "_pose.txt");
 
     Eigen::Matrix4f pose_mat;
     pose_mat.setIdentity (4, 4);
 
-    std::stringstream filestream_pose;
-    filestream_pose << data_dir << "/" << pose_file;
-    pose_file = filestream_pose.str ();
+    pose_file = data_dir + "/" + pose_file;
```
Comment (SergioRAgostinho) [inline recognition/src/face_detection/face_detector_data_provider.cpp]: Same comment here.

File: recognition/src/face_detection/face_detector_data_provider.cpp (hunk 13 above)
Comment (SunBlack) [reply recognition/src/face_detection/face_detector_data_provider.cpp]: Not sure if there is a warning, but I didn't saw one until now. And btw: If I search for regex `"."`  in all `*.cpp;*.h;*.hpp;*.qrc;*.cu` files I have 2119 hit - so I believe there is no check.

**Pull request merged**